### PR TITLE
Allow for overriding image upload path

### PR DESCRIPTION
**Pull request by willcodefortea**
Previously I've worked on a project that has quite a few images being uploaded (it's currently using ~ 28,000 images, and then that's multiplied for the renditions). In the not too distant future the file system will stop being able to parse the tree (in a reasonable time frame), and backups will no longer be able to taken!

This allows the user to override the default image file path with a more friendly version - the example I've given should scale for most systems. An added benefit for this type of file positioning is that the path is generated based on the contents, so files that share the same name will probably* have different paths.

*As I'm only using the first three characters of the checksum, there's still a chance that images sharing the same name will be positioned in the same folder, all be it pretty small.

### CODE REVIEW

**Comment by davecranwell** [comment]
So this is basically a sharded file system? Interesting.

**Comment by benemery** [comment]
Essentially yes. I've kept the original filename in the example I've given so URLs are still usable. But this mainly allows for the user to do whatever they like.

The same will need to be applied to `AbstractRendition` , and if you're happy with this then I'll gladly supply another pull request.

**Comment by kaedroho** [comment]
Thanks @benemery 
Could we change this so it hashes the filename instead? That would run much quicker

**Comment by benemery** [comment]
It would be very much quicker, you're right. But the issue there then is that files with the same name end up being in the same path. Hashing the content kills a few birds with one stone, but if the hash time is that much of an issue then it's easy to replace! (Although I'd probably argue that the time to upload the file in the first place will greatly outweigh the time spent hashing it.)

File: wagtail/wagtailimages/models.py [hunk 1]
```diff
@@ -71,6 +61,20 @@ class AbstractImage(models.Model, TagSearchable):
     focal_point_width = models.PositiveIntegerField(null=True, blank=True)
     focal_point_height = models.PositiveIntegerField(null=True, blank=True)
 
+    @staticmethod
+    def get_upload_to(instance, filename):
```
Comment (kaedroho) [inline wagtail/wagtailimages/models.py]: Could this be turned into a normal method?

Comment (benemery) [inline wagtail/wagtailimages/models.py]: Don't see why not, I'll get something over to you in the next couple of days? (Bit busy at the moment!)

**Comment by kinghelix** [comment]
Just wondering how this issue/request is tracking - it's a big requirement for us, so if there was a way it can make it in to 1.3, it would be awesome. Thanks!

**Comment by gasman** [comment]
@kinghelix It looks to me like this is just waiting for someone to change `get_upload_to` to a standard instance method - is that correct, @kaedroho?

**Comment by kaedroho** [comment]
@gasman Yep, it looks like that's the case

**Comment by willcodefortea** [comment]
Hi, sorry completely lost track of this. IIRC this was a little bit fiddly as various migrations require updating. I'll take a look this evening if you're willing to wait an additional day?

**Comment by willcodefortea** [comment]
That should do it. It's annoying to have that proxy method but that's the only way I could see to do it without introducing a classmethod instead. This way doesn't require any migrating tweaking though, which is nice.

**Comment by kinghelix** [comment]
Wow cool, this sounds great.

So my objective/hope with this feature is that we can group different "types" of images into different groups, to make the Content Team's job of choosing images easier - eg when they go to select a Staff Headshot, they only see the directory of previously uploaded headshots, not the page headers, side bar images etc etc.

Thanks!

**Comment by gasman** [comment]
@kinghelix What you're describing is very different from this feature (this is about automatically organising the files on disk, not a user-facing organisation scheme) - but you may be interested in the 'Collections' feature (#1751) which is also scheduled for 1.3.

**Comment by kinghelix** [comment]
@gasman 

Ahh, ok cool thanks. I had been directed here (I did think it sounded a little different to what I was after...) 

I had found the "collections" ticket & RFC a few days ago too - sounds great, excited to see it!

thanks everyone

cheers

**Comment by gasman** [comment]
Rebased on master: https://github.com/gasman/wagtail/tree/willcodefortea-feature/allow-upload-to-overriding

**Comment by gasman** [comment]
Merged in beb2b927d75767ac104d581665b1853ed4f714a5 - thanks!

**Pull request closed without merging**